File: src/Poker/game/HoldEm.java
package Poker.game;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

import Poker.hand.Hand;

public class HoldEm
{
   private ArrayList<Player> Players;
   private HoldEmState state;
   private boolean isSimulation;
   private int dealer;
   private Pots pots;
   private Deck deck;
   private Board board;
   private ActionLog actionLog;
   private ArrayList<HoldEmEvents> eventListeners = new ArrayList<HoldEmEvents>();
   
   public HoldEm(ArrayList<Player> Players, boolean isSimulation, Deck deck)
   {
      this.isSimulation = isSimulation;
      this.Players = Players;
      
      this.deck = deck;
      board = new Board();
      actionLog = new ActionLog();
      
      dealer = Players.size() - 1;
   }
   
   public void addEventListener(HoldEmEvents listener)
   {
      eventListeners.add(listener);
   }
   
   private void emitEvent(String method)
   {
      for (HoldEmEvents listener : eventListeners)
      {
         try
         {
            Method toInvoke = HoldEmEvents.class.getMethod(method);
            toInvoke.invoke(listener);
         }
         catch (Exception x)
         {
            // TODO: should this do anything?
         }
      }
   }
   
   public List<ActionLogEntry> getLogEntries()
   {
      return actionLog.getLog();
   }
   
   public GameState startHand()
   {
      deck.shuffle();
      board.clear();
      
      List<Player> players = getPlayersForMainPot();
      if (players.size() < 2)
      {
         return null;
      }
      pots = new Pots(players, Chip.BIG_BLIND, state);
      
      for (Player player : Players)
      {
         player.Fold();
      }
      
      if (isSimulation())
      {
         state = HoldEmState.DEAL_HOLES;
      }
      else
      {
         state = HoldEmState.BLINDS;
         dealCards();
      }
      
      pots.startRound(state);
      
      if (!changeDealer())
      {
         return null;
      }
      actionLog.addEntry("Started hand");
      
      return generateNextAction();
   }
   
   private boolean changeDealer()
   {
      return (dealer = getNextPlayerAtTable(dealer)) >= 0;
   }
   
   private List<Player> getPlayersForMainPot()
   {
      List<Player> players = new ArrayList<Player>();
      for (int i = getNextPlayerAtTable(dealer); i != dealer; i = getNextPlayerAtTable(i))
      {
         players.add(getPlayer(i));
      }
      players.add(getPlayer(dealer));
      return players;
   }
   
   public String getNextAction()
   {
//      if (potsRight)
//      {
//         return "";
//      }
//      Player player = getActionPlayer();
//      int call = getCall();
//      int minRaise = getMinRaise();
//      StringBuilder action = new StringBuilder(player.getName() + "\n");
//      int currBet = pots.getCurrentBet();
//      switch (state)
//      {
//      case BLINDS:
//         if (!gotSB)
//         {
//            if (call == getSmallBlind())
//            {
//               action.append("Post Small Blind Of " + call + "?");
//            }
//            else
//            {
//               action.append("Post " + call + " Toward Small Blind (All In)?");
//            }
//         }
//         else if (call == getBigBlind())
//         {
//            action.append("Post Big Blind Of " + call + "?");
//         }
//         else
//         {
//            action.append("Post " + call + " Toward Big Blind (All In)?");
//         }
//         break;
//      case BET_PREFLOP:
//         if (currBet == bigBlind && actionOn == bigBlindOn && minRaise > 0)
//         {
//            action.append("Option: Check or Raise?");
//            break;
//         }
//      case BET_FLOP:
//      case BET_TURN:
//      case BET_RIVER:
//         if (currBet == 0 && minRaise > 0)
//         {
//            action.append("Check or Open?");
//         }
//         else if (minRaise > 0)
//         {
//            action.append("Call " + call + ", Raise, or Fold?");
//         }
//         else
//         {
//            action.append("Call " + call + " (All In) or Fold?");
//         }
//         break;
//      default:
//         return "";
//      }
//      return action.toString();
      return "";
   }
   
   public GameState generateNextAction()
   {
      boolean isBettingRound = false;
      switch (state)
      {
         case BLINDS:
         case BET_PREFLOP:
         case BET_FLOP:
         case BET_TURN:
         case BET_RIVER:
            isBettingRound = true;
            break;
         default:
      }
      
      return isBettingRound ? new GameState(state, getActionPlayer(), getCall(), getMinRaise(), getMaxRaise(), getTotalPotSize(), getPlayersSimple()) : new GameState(state, getTotalPotSize(), getPlayersSimple());
   }
   
   private List<PlayerSimple> getPlayersSimple()
   {
      List<PlayerSimple> playersSimple = new ArrayList<PlayerSimple>();
      
      for (Player player : Players)
      {
         PlayerSimple playerSimple = new PlayerSimple(player.getName(), player.getChips(), player.hasHoleCards(), pots.getChipsThisRound(player));
         playersSimple.add(playerSimple);
      }
      
      return playersSimple;
   }
   
   public GameState Bet(int bet) throws Exception
   {
      actionLog.addEntry(pots.addToPot(bet));
      if (pots.isPotEven() && !pots.isHandOver())
      {
         moveState();
      }
      
      return generateNextAction();
   }
   
   public GameState Check() throws Exception
   {
      return Bet(0);
   }
   
   public GameState Fold() throws Exception
   {
      actionLog.addEntries(pots.fold());
      if (pots.isPotEven() && !pots.isHandOver())
      {
         moveState();
      }
      else if (pots.isHandOver())
      {
         state = HoldEmState.WINNER;
         Pot wonMainPot = pots.awardPot(getBoard());
         List<Player> winners = wonMainPot.getWinners(getBoard());
         ActionLogEntry entry;
         if (winners.size() == 1)
         {
            entry = new ActionLogEntry(winners.get(0), "won pot of " + wonMainPot.getSize());
         }
         else
         {
            entry = new ActionLogEntry(winners, "split pot of " + wonMainPot.getSize());
         }
         
         actionLog.addEntry(entry);
      }
      
      return generateNextAction();
   }
   
   private void moveState()
   {
      if (!isBettingOver())
      {
         switch (state)
         {
            case BLINDS:
            case DEAL_HOLES:
               state = HoldEmState.BET_PREFLOP;
               break;
            case BET_PREFLOP:
            case DEAL_FLOP:
               state = HoldEmState.BET_FLOP;
               break;
            case BET_FLOP:
            case DEAL_TURN:
               state = HoldEmState.BET_TURN;
               break;
            case BET_TURN:
            case DEAL_RIVER:
               state = HoldEmState.BET_RIVER;
               break;
            case BET_RIVER:
               state = HoldEmState.WINNER;
               break;
            case WINNER:
               state = state.getNextState();
               break;
            default:
               // TODO: throw exception here
         }
      }
      else
      {
         state = HoldEmState.WINNER;
      }
      
      pots.startRound(state);
   }
   
   // TODO: this will be only used by simulation
   private void changeState()
   {
      if (!isBettingOver())
      {
         state = state.getNextState();
      }
      else
      {
         switch (state)
         {
            case BLINDS:
               state = HoldEmState.DEAL_HOLES;
               break;
            case DEAL_HOLES:
            case BET_PREFLOP:
               state = HoldEmState.DEAL_FLOP;
               break;
            case DEAL_FLOP:
            case BET_FLOP:
               state = HoldEmState.DEAL_TURN;
               break;
            case DEAL_TURN:
            case BET_TURN:
               state = HoldEmState.DEAL_RIVER;
               break;
            case DEAL_RIVER:
            case BET_RIVER:
               state = HoldEmState.WINNER;
               break;
            case WINNER:
               state = state.getNextState();
               break;
            default:
               // TODO: throw exception here
         }
      }
   }
   
   private void dealCards()
   {
      List<Player> players = pots.getMainPot().getPlayers();
      for (Player player : players)
      {
         Card cards[] = new Card[2];
         for (int j = 0; j < 2; j++)
         {
            cards[j] = deck.dealCard();
         }
         player.DealHoleCards(cards);
      }
      
      Card[] cards = new Card[5];
      for (int i = 0; i < 5; i++)
      {
         cards[i] = deck.dealCard();
      }
      board.addBoard(cards);
   }
   
   // TODO: this will be only used by simulation
   public GameState deal()
   {
      switch (state)
      {
         case DEAL_HOLES:
            dealHoles();
            break;
         case DEAL_FLOP:
            dealFlop();
            break;
         case DEAL_TURN:
            dealTurn();
            break;
         case DEAL_RIVER:
            dealRiver();
            break;
         default:
            startHand();
            if (isSimulation())
            {
               deal();
            }
            return generateNextAction();
      }
      
      changeState();
      if (isSimulation)
      {
         changeState();
      }
      pots.startRound(state);
      
      return generateNextAction();
   }
   
   // TODO: this will be only used by simulation
   private void dealHoles()
   {
      List<Player> players = pots.getMainPot().getPlayers();
      for (Player player : players)
      {
         Card cards[] = new Card[2];
         for (int j = 0; j < 2; j++)
         {
            cards[j] = deck.dealCard();
         }
         player.DealHoleCards(cards);
      }
      emitEvent("holesDealt");
      actionLog.addEntry("Dealt hole cards");
   }
   
   // TODO: this will be only used by simulation
   private void dealFlop()
   {
      Card[] cards = new Card[3];
      for (int i = 0; i < 3; i++)
      {
         cards[i] = deck.dealCard();
      }
      board.addFlop(cards);
      emitEvent("flopDealt");
      actionLog.addEntry("Dealt flop");
   }
   
   // TODO: this will be only used by simulation
   private void dealTurn()
   {
      board.addTurn(deck.dealCard());
      emitEvent("turnRiverDealt");
      actionLog.addEntry("Dealt turn");
   }
   
   // TODO: this will be only used by simulation
   private void dealRiver()
   {
      board.addRiver(deck.dealCard());
      emitEvent("turnRiverDealt");
      actionLog.addEntry("Dealt river");
   }
   
   public Card[] getFlop()
   {
      return board.getFlop();
   }
   
   public Card getTurn()
   {
      return board.getTurn();
   }
   
   public Card getRiver()
   {
      return board.getRiver();
   }
   
   public Card[] getBoard()
   {
      Card[] cards;
      switch (state)
      {
         case BLINDS:
         case DEAL_HOLES:
         case BET_PREFLOP:
            return null;
         case DEAL_FLOP:
         case BET_FLOP:
            return board.getFlop();
         case DEAL_TURN:
         case BET_TURN:
            cards = new Card[4];
            System.arraycopy(board.getFlop(), 0, cards, 0, 3);
            cards[3] = board.getTurn();
            return cards;
         default:
            cards = new Card[5];
            System.arraycopy(board.getFlop(), 0, cards, 0, 3);
            cards[3] = board.getTurn();
            cards[4] = board.getRiver();
            return cards;
      }
   }
   
   public boolean isSimulation()
   {
      return isSimulation;
   }
   
   private int getCall()
   {
      switch (state)
      {
         case DEAL_HOLES:
         case DEAL_FLOP:
         case DEAL_TURN:
         case DEAL_RIVER:
         case WINNER:
            return 0;
         default:
            // TODO: throw exception here maybe?
      }
      
      int call = 0;
      try
      {
         call = pots.getCall(); 
      }
      catch (Exception x)
      {
         
      }
      
      return call;
   }
   
   private int getMinRaise()
   {
      switch (state)
      {
         case BLINDS:
         case DEAL_HOLES:
         case DEAL_FLOP:
         case DEAL_TURN:
         case DEAL_RIVER:
         case WINNER:
            return 0;
         default:
            // TODO: throw exception here maybe?
      }
      
      int minRaise = 0;
      try
      {
         minRaise = pots.getMinRaise(); 
      }
      catch (Exception x)
      {
         // TODO: ???
      }
      return minRaise;
   }
   
   private int getMaxRaise()
   {
      switch (state)
      {
         case BLINDS:
         case DEAL_HOLES:
         case DEAL_FLOP:
         case DEAL_TURN:
         case DEAL_RIVER:
         case WINNER:
            return 0;
         default:
            // TODO: Throw exception here maybe?
      }
      
      int maxRaise = 0;
      try
      {
         maxRaise = pots.getMaxRaise(); 
      }
      catch (Exception x)
      {
         // TODO: ???
      }
      return maxRaise;
   }
   
   public Player getDealerPlayer()
   {
      return getPlayer(dealer);
   }
   
   private Player getActionPlayer()
   {
      return pots.getNextActionPlayer();
   }
   
   public int getPlayersCount()
   {
      return Players.size();
   }
   
   public Player getPlayer(int i)
   {
      return Players.get(i);
   }
   
   public boolean isBettingOver()
   {
      return pots.isBettingOver();
   }
   
   public String potsToString()
   {
      return pots.toString();
   }
   
   public Pot getMainPot()
   {
      return pots.getMainPot();
   }
   
   public int getChipsThisRound(Player player)
   {
      return pots.getChipsThisRound(player);
   }
   
   public Pot awardPot()
   {
      return pots.awardPot(getBoard());
   }
   
   public boolean isMainAwarded()
   {
      //return !pots.hasMainPot();
      return false;
   }
   
   public int getTotalPotSize()
   {
      return pots.getTotalSize();
   }
   
   private int getNextPlayerAtTable(int i) throws IllegalArgumentException
   {
      int players = getPlayersCount();
      if (i >= players)
      {
         throw new IllegalArgumentException();
      }
      
      int start = i;
      do
      {
         i = (i + 1) % players;
         if (i == start)
         {
            return -1;
         }
      } while (getPlayer(i) == null);
      return i;
   }
   
   public Hand getWinningHand()
   {
      return getMainPot().getBestHand(getBoard());
   }
   
   public List<Player> getWinners()
   {
      return getMainPot().getWinners(getBoard());
   }
}

File: server/game/holdEm.js
let Board = require('./board');
let ActionLog = require('./actionLog');
let ActionLogEntry = require('./actionLogEntry');
let PlayerClient = require('./playerClient');
let HoldEmState = require('./holdEmState');
let GameStateClient = require('./gameStateClient');
let NextAction = require('./nextAction');
let Pots = require('./pots');

let HoldEm = (maxPlayers, bigBlind, deck, autoPostBlinds = false) =>
{
	let players = Array(maxPlayers).fill(null);
	let newPlayers = [];
	let pendingPlayers = [];
	
	let dealerIndex = maxPlayers - 1;
	let pots;
	let board;
	let actionLog = ActionLog();
	
	let gameState = {};
	let state = HoldEmState().NO_GAME;
	let winners = null;
	
	let lastAction = [];
	
	let addLastAction = (action) =>
	{
		lastAction.push(action);
	}
	
	let setLastAction = (action) =>
	{
		lastAction = [];
		addLastAction(action);
	}
	
	let getNumPlayersAtTable = () =>
	{
		let count = 0;
		players.forEach((player) =>
		{
			if (player)
			{
				count++;
			}
		});
		return count;
	}
	
	let getNextPlayerAtTableNotNew = (i) =>
	{
		let playersCount = players.length;
		if (i >= playersCount)
		{
			throw 'Invalid index ' + i;
		}
		
		let nextIndex = (i + 1) % playersCount;
		let index = nextIndex;
		while (!players[index] || index !== i && (newPlayers.indexOf(players[index]) !== -1 || pendingPlayers.indexOf(players[index]) !== -1))
		{
			index = (index + 1) % playersCount;
		}
		
		if (index == i)
		{
			// Looped back to original player; let all new players in for free
			newPlayers = [];
			
			index = nextIndex;
			while (!players[index] || index !== i && (newPlayers.indexOf(players[index]) !== -1 || pendingPlayers.indexOf(players[index]) !== -1))
			{
				index = (index + 1) % playersCount;
			}
			
			return index;
		}
		return index;
	}
	
	let getNextIndexAtTable = (i) =>
	{
		let playersCount = players.length;
		if (i >= playersCount)
		{
			throw 'Invalid index' + i;
		}
		
		let index = (i + 1) % playersCount;
		while (!players[index])
		{
			index = (index + 1) % playersCount;
		}
		
		return index;
	}
	
	let getPlayersForMainPot = () =>
	{
		let mainPlayers = [];
		for (let i = getNextIndexAtTable(dealerIndex); i != dealerIndex; i = getNextIndexAtTable(i))
		{
			mainPlayers.push(players[i]);
		}
		mainPlayers.push(players[dealerIndex]);
		return mainPlayers;
	}
	
	let getOngoingRoundAction = (player) => pots ? pots.getOngoingActionThisRound(player) : null;
	
	let getPlayerClients = (nextActionPlayer) =>
	{
		let playerClients = [];
		for (let i = 0; i < players.length; i++)
		{
			let player = players[i];
			
			let isDealer = dealerIndex === i;
			let playerClient;
			if (player)
			{
				playerClient = PlayerClient(isDealer, player, player.hasHoleCards(), getOngoingRoundAction(player), player === nextActionPlayer);
			}
			else
			{
				playerClient = PlayerClient(isDealer);
			}
			playerClients.push(playerClient);
		}
		return playerClients;
	}
	
	let moveState = () =>
	{
		if (pots.isEven() && pots.hasTwoOrMorePlayers())
		{
			if (!pots.isBettingOver())
			{
				switch (state)
				{
					case HoldEmState().BLINDS:
						actionLog.addSystemEntry('Dealt holes');
						state = HoldEmState().BET_PREFLOP;
						dealCards();
						break;
					case HoldEmState().BET_PREFLOP:
						actionLog.addSystemEntry('Dealt flop');
						state = HoldEmState().BET_FLOP;
						dealCards();
						break;
					case HoldEmState().BET_FLOP:
						actionLog.addSystemEntry('Dealt turn');
						state = HoldEmState().BET_TURN;
						dealCards();
						break;
					case HoldEmState().BET_TURN:
						actionLog.addSystemEntry('Dealt river');
						state = HoldEmState().BET_RIVER;
						dealCards();
						break;
					case HoldEmState().BET_RIVER:
						state = HoldEmState().WINNER;
						break;
					case HoldEmState().WINNER:
					case HoldEmState().NO_GAME:
						state = HoldEmState().BLINDS;
						break;
				}
			}
			else
			{
				dealRemainingCards();
				state = HoldEmState().WINNER;
			}
			
			pots.startRound(state);
		}
		else if (!pots.hasTwoOrMorePlayers())
		{
			state = HoldEmState().WINNER;
			
			pots.startRound(state);
		}
	}
	
	let dealCards = () =>
	{
		switch (state)
		{
			case HoldEmState().BET_PREFLOP:
				dealHoles();
				break;
			case HoldEmState().BET_FLOP:
				dealFlop();
				break;
			case HoldEmState().BET_TURN:
			case HoldEmState().BET_RIVER:
				dealTurnRiver();
				break;
		}
	}
	
	let dealRemainingCards = () =>
	{
		switch (state)
		{
			case HoldEmState().BET_PREFLOP:
				dealHoles();
			case HoldEmState().BET_FLOP:
				dealFlop();
			case HoldEmState().BET_TURN:
				dealTurnRiver();
			case HoldEmState().BET_RIVER:
				dealTurnRiver();
				break;
		}
	}
	
	let dealHoles = () =>
	{
		let mainPlayers = pots.getMainPot().getPlayers();
		mainPlayers.forEach((mainPlayer) =>
		{
			mainPlayer.deal(deck.dealCards(2));
		});
		addLastAction('dealHoles');
	}
	
	let dealFlop = () =>
	{
		board.addFlop(deck.dealCards(3));
		addLastAction('dealFlop');
	}
	
	let dealTurnRiver = () =>
	{
		board.addTurnRiver(deck.deal());
		addLastAction('dealTurnRiver');
	}
	
	let getCall = () =>
	{
		switch (state)
		{
			case HoldEmState().WINNER:
			case HoldEmState().NO_GAME:
				return 0;
		}
		
		return pots.getCall();
	}
	
	let getMinRaise = () =>
	{
		switch (state)
		{
			case HoldEmState().BLINDS:
			case HoldEmState().WINNER:
			case HoldEmState().NO_GAME:
				return 0;
		}
		
		return pots.getMinRaise();
	}
	
	let getMaxRaise = () =>
	{
		switch (state)
		{
			case HoldEmState().BLINDS:
			case HoldEmState().WINNER:
			case HoldEmState().NO_GAME:
				return 0;
		}
		
		return pots.getMaxRaise();
	}
	
	let getBoard = () => board ? board.getBoard() : null;
	
	let potPlayersToShowdownPlayers = (potPlayers, potWinners, lastAggressor, foldedPlayers) =>
	{
		let playerClients = Array(players.length).fill(null);
		let i;
		let startedLoop = false;
		
		let originIndex = players.indexOf(lastAggressor);
		if (originIndex === -1)
		{
			originIndex = (dealerIndex + 1) % players.length;
		}
		
		for (i = originIndex; i != originIndex || !startedLoop; i = (i + 1) % players.length)
		{
			startedLoop = true;
			let player = players[i];
			let isDealer = dealerIndex === i;
			
			let playerClient;
			if (player)
			{
				let newPlayer = player.clone();
				if (foldedPlayers.indexOf(player) !== -1)
				{
					newPlayer.fold();
					playerClient = PlayerClient(isDealer, newPlayer, false);
				}
				else
				{
					let hasHoleCards = newPlayer.hasHoleCards();
					if (potPlayers.indexOf(player) === -1)
					{
						newPlayer.fold();
					}
					
					if (potPlayers.length == 1)
					{
						newPlayer.fold();
					}
					
					playerClient = PlayerClient(isDealer, newPlayer, hasHoleCards);
				}
			}
			else
			{
				playerClient = PlayerClient(isDealer);
			}
			
			playerClients[i] = playerClient;
			
			if (player && potWinners.find((winner) => winner.getName() === player.getName()))
			{
				break;
			}
			else if (player && potPlayers.indexOf(player) !== -1 && potWinners.indexOf(player) === -1 && foldedPlayers.indexOf(player) === -1)
			{
				foldedPlayers.push(player);
			}
		}
		
		for (i = (i + 1) % players.length; i != originIndex; i = (i + 1) % players.length)
		{
			let player = players[i];
			let isDealer = dealerIndex === i;
			
			let playerClient;
			if (player && potWinners.find((winner) => winner.getName() === player.getName()))
			{
				let newPlayer = player.clone();
				playerClient = PlayerClient(isDealer, newPlayer, newPlayer.hasHoleCards());
			}
			else if (player)
			{
				let newPlayer = player.clone();
				let hasHoleCards = newPlayer.hasHoleCards();
				if (foldedPlayers.indexOf(player) !== -1)
				{
					hasHoleCards = false;
				}
				
				if (potPlayers.indexOf(player) !== -1 && foldedPlayers.indexOf(player) === -1)
				{
					foldedPlayers.push(player);
				}
				
				newPlayer.fold();
				
				playerClient = PlayerClient(isDealer, newPlayer, hasHoleCards);
			}
			else
			{
				playerClient = PlayerClient(isDealer);
			}
			playerClients[i] = playerClient;
		}
		
		return playerClients;
	}
	
	let awardPots = () =>
	{
		if (!winners)
		{
			winners = [];
			let awardedPots = pots.awardPots(getBoard());
			
			let foldedPlayers = [];
			awardedPots.pots.forEach((pot) =>
			{
				let action;
				if (pot.winnerState.winners.length == 1)
				{
					action = "won pot of " + pot.potSize;
				}
				else
				{
					action = "split pot of " + pot.potSize;
				}
				let entry = ActionLogEntry(action, pot.winnerState.winners);
				actionLog.addEntry(entry);
				
				winners.push({players: potPlayersToShowdownPlayers(pot.players, pot.winnerState.winners, awardedPots.lastAggressor, foldedPlayers), hand: pot.winnerState, potSize: pot.potSize});
			});
		}
	}
	
	let generateGameState = () =>
	{
		let nextAction = null;
		let nextActionPlayer = null;
		switch (state)
		{
			case HoldEmState().BLINDS:
			case HoldEmState().BET_PREFLOP:
			case HoldEmState().BET_FLOP:
			case HoldEmState().BET_TURN:
			case HoldEmState().BET_RIVER:
				nextAction = NextAction(getCall(), getMinRaise(), getMaxRaise());
				nextActionPlayer = pots.getNextActionPlayer();
				break;
			case HoldEmState().WINNER:
				awardPots();
				break;
			default:
		}
		
		gameState = GameStateClient(getPlayerClients(nextActionPlayer), bigBlind, nextAction, nextActionPlayer, self.getPotSizeWithoutRound(), getBoard(), null, lastAction);
		lastAction = [];
	}
	
	let self =
	{
		getLogEntries: () => actionLog.getEntries(),
		getTotalPotSize: () => pots ? pots.getTotalSize() : 0,
		getPotSizeWithoutRound: () => pots ? pots.getSizeWithoutRound() : 0,
		getState: () => state,
		getGameState: () => gameState.clone(),
		getWinners: () => winners,
		startHand: () =>
		{
			deck.shuffle();
			
			// TODO: removing players
			// players.forEach((player) =>
			// {
			// 	if (player && player.getChips === 0)
			// 	{
			// 		self.removePlayer(player);
			// 	}
			// });
			
			dealerIndex = getNextPlayerAtTableNotNew(dealerIndex);
			
			pendingPlayers = [];
			
			board = Board();
			
			let mainPlayers = getPlayersForMainPot();
			if (mainPlayers.length < 2)
			{
				return null;
			}
			pots = Pots(mainPlayers, newPlayers, bigBlind);
			
			newPlayers = [];
			
			players.forEach((player) =>
			{
				if (player)
				{
					player.fold();
				}
			});
			
			state = HoldEmState().BLINDS;
			winners = null;
			
			pots.startRound(state);
			
			actionLog.addSystemEntry("Started hand");
			
			generateGameState();
			
			if (autoPostBlinds)
			{
				while (state === HoldEmState().BLINDS)
				{
					self.call(true);
				}
			}
		},
		setToNoGame: () =>
		{
			if (state === HoldEmState().WINNER)
			{
				state = HoldEmState().NO_GAME;
				winners = null;
				
				players.forEach((player) =>
				{
					if (player)
					{
						player.fold();
					}
				});
				
				generateGameState();
			}
		},
		bet: (addition) =>
		{
			let newGameState = self.getGameState();
			let nextAction = newGameState.nextAction;
			let nextActionPlayer = newGameState.nextActionPlayer;
			if (addition < nextAction.call && nextActionPlayer.getChips() > 0)
			{
				throw 'Bet ' + addition + ' is less than the call ' + nextAction.call;
			}
			
			if (addition > nextAction.call && addition < nextAction.minRaise)
			{
				throw 'Bet ' + addition + ' is less than the min raise ' + nextAction.minRaise;
			}
			
			if (addition !== nextAction.call && addition > nextAction.maxRaise)
			{
				addition = nextAction.maxRaise;
			}
			
			if (addition > nextAction.call)
			{
				setLastAction('raise');
			}
			
			actionLog.addEntry(pots.addToPot(addition));
			moveState();
			generateGameState();
		},
		call: (autoPosted = false) =>
		{
			if (!autoPosted)
			{
				setLastAction('call');
			}
			self.bet(self.getGameState().nextAction.call);
		},
		check: () =>
		{
			setLastAction('check');
			self.bet(0);
		},
		fold: () =>
		{
			actionLog.addEntries(pots.fold());
			setLastAction('fold');
			moveState();
			generateGameState();
		},
		foldOutOfTurn: (player) =>
		{
			if (state !== HoldEmState().NO_GAME && state !== HoldEmState().WINNER)
			{
				actionLog.addEntries(pots.foldOutOfTurn(player));
				setLastAction('fold');
				moveState();
			}
			else if (state === HoldEmState().WINNER)
			{
				if (getNumPlayersAtTable() === 0)
				{
					self.setToNoGame();
				}
			}
			generateGameState();
		},
		removePlayer: (player) =>
		{
			let index = players.indexOf(player);
			if (index === -1)
			{
				throw "Can't remove player that doesn't exist";
			}
			
			players[index] = null;
			
			self.foldOutOfTurn(player);
		},
		addPlayer: (player, i) =>
		{
			if (players.indexOf(player) !== -1)
			{
				throw 'Tried to add player that already exists';
			}
			
			if (players[i])
			{
				throw 'Tried to add player at occupied position ' + i;
			}
			
			if (state === HoldEmState().WINNER || state === HoldEmState().NO_GAME)
			{
				newPlayers.push(player);
			}
			else
			{
				pendingPlayers.push(player);
			}
			
			players[i] = player;
			
			generateGameState();
		}
	}
	
	generateGameState();
	
	return self;
}

module.exports = HoldEm;
